### PR TITLE
[FLINK-14672][sql-client] Make Executor stateful in sql client

### PR DESCRIPTION
**Pull request by openinx**
## What is the purpose of the change

Now we only have the embedded mode SQL to interact with Flink cluster, will support the multi-statement and DDL such create table, which means we need to store some state for the same session. On the other hande, it will support the gateway mode in future, means it will maitain multi-sessions at the server side. 

So it's necessary to be stateful for the Executor (both LocalExecutor and GatewayExecutor). 

The pull request provide two extra API in Executor interface to manage session life-cycle:  
1. String openSession(SessionContext context);  it will create a new session based on the passed session context, the session will have the ExecutionContext/TableEnviroment/UDF informations (etc).  and the openSession will return a uuid string which represent the session identification, the following operations of this session will attach the sessionId to request the Executor (both local and gateway). 
2. String closeSession(String sessionId);   it will close the session(clear the session related resources) with given session identify.

For each session, we will openSesion firstly, then execute some DDL or DML, and closeSession finally. When executing the DDL or DML, it will share the same session context and execution context for the same session identifier.

## Brief change log

- 4b5c6f4491 [FLINK-14672][sql-client] Make Executor stateful in sql client 
- e60ea8bf25 [FLINK-14672][sql-client] Simplify the executor backend state for each session: The SessionContext is only used for frontend user to open session, all the backend informations are maintained in ExecutionContext.

## Verifying this change

The LocalExecutorITCase addressed all the cases.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by openinx** [comment]
FYI @twalthr @KurtYoung , please help the reviewing if have time. Thanks.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 1]
```diff
@@ -171,12 +175,19 @@ public LocalExecutor(URL defaultEnv, List<URL> jars, List<URL> libraries) {
...
 	 * Constructor for testing purposes.
 	 */
-	public LocalExecutor(Environment defaultEnvironment, List<URL> dependencies, Configuration flinkConfig, CustomCommandLine commandLine, ClusterClientServiceLoader clusterClientServiceLoader) {
+	public LocalExecutor(
+		Environment defaultEnvironment,
+		List<URL> dependencies,
+		Configuration flinkConfig,
+		CustomCommandLine commandLine,
+		ClusterClientServiceLoader clusterClientServiceLoader
+	) {
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: nit: align with other code format

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 2]
```diff
@@ -188,10 +199,76 @@ public void start() {
...
+
 	@Override
-	public Map<String, String> getSessionProperties(SessionContext session) throws SqlExecutionException {
-		final Environment env = getOrCreateExecutionContext(session)
-			.getMergedEnvironment();
+	public String openSession(SessionContext sessionContext) throws SqlExecutionException {
+		String sessionId = sessionContext.getSessionId();
+		ExecutionContext previousContext = this.contextMap.putIfAbsent(sessionId, createExecutionContext(sessionContext));
+		if (previousContext != null) {
+			throw new SqlExecutionException("Found another session with the same session identifier: " + sessionContext.getSessionId());
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: `sessionContext.getSessionId()` -> `sessionId`

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 3]
```diff
@@ -188,10 +199,76 @@ public void start() {
...
+		}
+		return sessionContext.getSessionId();
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: `sessionContext.getSessionId()` -> `sessionId`

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 4]
```diff
@@ -200,69 +277,98 @@ public void start() {
 	}
 
 	@Override
-	public List<String> listCatalogs(SessionContext session) throws SqlExecutionException {
-		final ExecutionContext<?> context = getOrCreateExecutionContext(session);
+	public void resetSessionProperties(String sessionId) throws SqlExecutionException {
+		// Renew the ExecutionContext by using the default environment.
+		this.contextMap.put(sessionId, createExecutionContext(defaultEnvironment));
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: reset session property should keep the content which read from session environment file? Now it looks we also git rid of this.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 5]
```diff
@@ -200,69 +277,98 @@ public void start() {
...
 
-		final TableEnvironment tableEnv = context
-			.createEnvironmentInstance()
-			.getTableEnvironment();
+	@Override
+	public void setSessionProperty(String sessionId, String key, String value) throws SqlExecutionException {
+		Environment env = getExecutionContext(sessionId).getEnvironment();
+		Environment newEnv = Environment.enrich(env, ImmutableMap.of(key, value), ImmutableMap.of());
+		// Renew the ExecutionContext by merging the default environment and new environment.
+		this.contextMap.put(sessionId, createExecutionContext(defaultEnvironment, newEnv));
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: I think we can directly call `createExecutionContext(newEnv)` here, `defaultEnvironment` should be already covered by the `env` you got from current execution context

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 6]
```diff
@@ -200,69 +277,98 @@ public void start() {
...
-		return context.wrapClassLoader(() -> Arrays.asList(tableEnv.listCatalogs()));
+	@Override
+	public void addView(String sessionId, String name, String query) throws SqlExecutionException {
+		Environment env = getExecutionContext(sessionId).getEnvironment();
+		Environment newEnv = Environment.enrich(
+			env,
+			ImmutableMap.of(),
+			ImmutableMap.of(name, ViewEntry.create(name, query)));
+		// Renew the ExecutionContext.
+		this.contextMap.put(sessionId, createExecutionContext(defaultEnvironment, newEnv));
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: ditto

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 7]
```diff
@@ -347,14 +441,16 @@ public String explainStatement(SessionContext session, String statement) throws
...
+		return executeQueryInternal(sessionId, context, query);
 	}
 
 	@Override
-	public TypedResult<List<Tuple2<Boolean, Row>>> retrieveResultChanges(SessionContext session,
-			String resultId) throws SqlExecutionException {
+	public TypedResult<List<Tuple2<Boolean, Row>>> retrieveResultChanges(
+		String sessionId,
+		String resultId
+	) throws SqlExecutionException {
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: nit: format

**Comment by TisonKun** [comment]
I think @aljoscha also works on this field. It is better to align works from merge one or the other.

**Comment by openinx** [comment]
Sync with @aljoscha on wednesday,  he should  know the issue I working.  Of course,  feel free to feedback if anything I missed, FYI @TisonKun , @aljoscha , Thanks.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ExecutionContext.java [hunk 8]
```diff
@@ -111,71 +113,56 @@
...
-	public ExecutionContext(Environment defaultEnvironment, SessionContext sessionContext, List<URL> dependencies,
-				Configuration flinkConfig, Options commandLineOptions, List<CustomCommandLine> availableCommandLines) throws FlinkException {
-		this(defaultEnvironment, sessionContext, dependencies, flinkConfig, new DefaultClusterClientServiceLoader(), commandLineOptions, availableCommandLines);
+	private TableEnvironment tableEnv;
+	private ExecutionEnvironment execEnv;
+	private StreamExecutionEnvironment streamExecEnv;
+	private Executor executor;
+
+	public ExecutionContext(
+		Environment environment,
```
[the remark this answers is not in the document]
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ExecutionContext.java]: I've set the checksytle in my ide, seems the Selection-format makes the indent happen again. Let me check.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ExecutionContext.java [hunk 9]
```diff
@@ -111,71 +113,56 @@
...
+		List<URL> dependencies,
+		Configuration flinkConfig,
+		Options commandLineOptions,
+		List<CustomCommandLine> availableCommandLines) throws FlinkException {
+		this(environment,
```
[the remark this answers is not in the document]
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ExecutionContext.java]: OK.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 1 above)
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: OK

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 2 above)
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: OK

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 3 above)
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: OK.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 4 above)
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: Yeah,  I noticed here.  I planed to not saving the session environment file in ExecutionContext before, so here seems no method to reset back to the session env.   Let me re-think about the  design , maybe we still need the session env..

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 5 above)
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: Sounds good, thanks for pointing it out.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 6 above)
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: Will also change this.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 7 above)
Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: OK.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ResultStore.java [hunk 10]
```diff
@@ -63,7 +63,7 @@ public ResultStore(Configuration flinkConfig) {
 
 		if (env.getExecution().inStreamingMode()) {
 			// determine gateway address (and port if possible)
-			final InetAddress gatewayAddress = getGatewayAddress(env.getDeployment());
+			InetAddress gatewayAddress = getGatewayAddress(env.getDeployment());
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ResultStore.java]: why deleting final?

Comment (openinx) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ResultStore.java]: I think it's unintentional change,  will revert it.

**Comment by twalthr** [comment]
Thanks for pinging me @KurtYoung. I would like to take a look.

**Comment by KurtYoung** [comment]
@twalthr do you have any comments on this? I will merge this if you don't mind.

**Comment by KurtYoung** [comment]
> A quick question: shall we introduce a type `SessionID` as wrapper of `String` instead of just using `String`? We can gain type safety & expressive code then.

We could revisit this when necessary, currently I don't see any strong reason to make it a `Class` instead of a `String`

**Comment by openinx** [comment]
bq. We could revisit this when necessary, currently I don't see any strong reason to make it a Class instead of a String
+1.

**Comment by KurtYoung** [comment]
merging this now

**Pull request closed without merging**